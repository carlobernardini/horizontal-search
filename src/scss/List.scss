.orizzonte__list {
    @extend %list;
    margin: 5px 0 0;
    min-width: $list-min-width;
    animation: listIn .1s linear;
    transform-origin: $list-transform-origin;

    &--right {
        left: auto;
        right: 0;
        transform-origin: top right;
    }

    &:hover {
        .orizzonte__list-controls--mouseover {
            margin: 5px 10px;
            transform: scale(1);
            max-height: 100px;
            opacity: 1;
            transition-delay: 0s;
        }
    }
}

.orizzonte__list-controls {
    display: flex;
    margin: 0 10px 5px;

    &--mouseover {
        max-height: 0;
        opacity: 0;
        transform: scale(.9);
        overflow: hidden;
        transition: all .3s;
        transition-delay: .5s;
    }
}

.orizzonte__list-control {
    outline: none;
    border: none;
    color: #fff;
    padding: 5px 0;
    cursor: pointer;
    transition: all .3s;
    display: table-cell;
    text-align: center;
    flex: 1;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    @extend %fonts;

    + .orizzonte__list-control {
        margin-left: 3px;
    }
}

.orizzonte__list-done {
    background-color: $btn-done-color;
    border-radius: 5px;
    flex-grow: 1;
    font: {
        weight: 500;
    }

    &:hover {
        background-color: lighten($btn-done-color, 5%);
    }
}

.orizzonte__list-remove {
    background-color: #FFF;
    color: $btn-remove-color;
    transition: color .3s;
}

.orizzonte__item {
    min-width: 175px;

    &--filters {
        + .orizzonte__item--filters {
            padding-top: 5px;
            margin-top: 5px;
            border-top: 1px solid #eee;
        }
    }
}

.orizzonte__item-clickable {
    @extend %list-item;
}
