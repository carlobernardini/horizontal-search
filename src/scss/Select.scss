@import 'base';

.orizzonte__filter-select {
    @extend %select;

    &:focus {
        border-color: $color-a;
    }

    &--disabled {
        @extend %select-disabled;
    }
}
