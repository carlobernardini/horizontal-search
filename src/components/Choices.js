import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
    includes, isEqual, uniqueId, without
} from 'lodash-es';
import { DISPLAY_NAME_FILTER_CHOICES, NAME_PREFIX_CHECKBOX, NAME_PREFIX_RADIO } from '../constants';
import CheckBox from './CheckBox';
import Caption from './Caption';
import FilterInfo from './FilterInfo';
import RadioButton from './RadioButton';
import '../scss/Filter.scss';

class Choices extends Component {
    renderChoices() {
        const {
            fieldName, multiple, onUpdate, options, value
        } = this.props;

        if (multiple) {
            return options.map((option, i) => (
                <CheckBox
                    key={ i }
                    id={ uniqueId(NAME_PREFIX_CHECKBOX) }
                    disabled={ option.disabled }
                    facetCount={ option.facetCount }
                    value={ option.value }
                    selected={ (value || []).indexOf(option.value) > -1 }
                    label={ option.label || option.value }
                    onChange={ (selected) => {
                        let newValue = (value || []).slice(0);
                        if (selected && !includes(newValue, option.value)) {
                            newValue.push(option.value);
                        }
                        if (!selected && includes(newValue, option.value)) {
                            newValue = without(newValue, option.value);
                        }
                        if (isEqual(newValue, value)) {
                            return false;
                        }
                        onUpdate(newValue.length ? newValue : null);
                        return true;
                    }}
                />
            ));            
        }

        return options.map((option, i) => (
            <RadioButton
                key={ i }
                id={ uniqueId(NAME_PREFIX_RADIO) }
                name={ fieldName }
                disabled={ option.disabled }
                facetCount={ option.facetCount }
                value={ option.value }
                selected={ value === option.value }
                label={ option.label || option.value }
                onChange={ (selectedValue) => (onUpdate(selectedValue)) }
            />
        ));
    }

    renderNoPreference() {
        const { noPreferenceLabel, multiple } = this.props;

        if (multiple || !noPreferenceLabel) {
            return null;
        }

        const { fieldName, onUpdate, value } = this.props;

        return (
            <RadioButton
                id={ uniqueId(NAME_PREFIX_RADIO) }
                name={ fieldName }
                value="no-preference"
                selected={ !value }
                label={ noPreferenceLabel }
                onChange={ () => (onUpdate(null)) }
            />
        );
    }

    render() {
        const { information, label } = this.props;
        return (
            <div
                className="orizzonte__filter"
            >
                <FilterInfo
                    information={ information }
                />
                <Caption>
                    { label }
                </Caption>
                { this.renderNoPreference() }
                { this.renderChoices() }
            </div>
        );
    }
}

Choices.displayName = DISPLAY_NAME_FILTER_CHOICES;

Choices.propTypes = {
    /** Field name for this filter, to be used in composed query */
    fieldName: PropTypes.string.isRequired,
    information: PropTypes.string,
    /** Label for this filter section */
    label: PropTypes.string,
    /** Whether to show checkboxes (true) or radios (false) */
    multiple: PropTypes.bool,
    /** Label to show if you want to include a 'no preference' option
        Only available for radio groups */
    noPreferenceLabel: PropTypes.string,
    /** Internal callback for when filter value has changed */
    onUpdate: PropTypes.func,
    /** List of selectable options (value is required) */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            disabled: PropTypes.bool,
            facetCount: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ]),
            label: PropTypes.string,
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired
        })
    ).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(
            PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ])
        )
    ])
};

Choices.defaultProps = {
    information: null,
    label: null,
    multiple: false,
    noPreferenceLabel: null,
    onUpdate: () => {},
    value: null
};

export default Choices;
