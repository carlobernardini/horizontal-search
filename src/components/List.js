import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/List.scss';

class List extends Component {
    renderDoneBtn() {
        const {
            doneBtn, doneBtnLabel, isFilterGroup, onApply
        } = this.props;

        if (!isFilterGroup || !doneBtn) {
            return null;
        }

        return (
            <li>
                <button
                    type="button"
                    className="orizzonte__list-done"
                    onClick={ onApply }
                >
                    { doneBtnLabel || 'Done' }
                </button>
            </li>
        );
    }

    renderItems() {
        const {
            values, items, isFilterGroup, onUpdate
        } = this.props;

        if (isFilterGroup) {
            return React.Children.map(items, (item, i) => (
                <li
                    className={ classNames('orizzonte__item', {
                        'orizzonte__item--filters': isFilterGroup
                    }) }
                    key={ i }
                >
                    { React.cloneElement(item, {
                        value: ((v, fn) => {
                            if (!(fn in v)) {
                                return null;
                            }
                            return v[fn];
                        })(values, item.props.fieldName),
                        onUpdate: (filterValue) => {
                            const { fieldName } = item.props;
                            onUpdate(fieldName, filterValue);
                        }
                    }) }
                </li>
            ));
        }

        return items.map((item, i) => (
            <li
                className={ classNames('orizzonte__item', {
                    'orizzonte__item--filters': isFilterGroup
                }) }
                key={ i }
            >
                { item }
            </li>
        ));
    }

    render() {
        const { orientation } = this.props;

        return (
            <ul
                className={ classNames('orizzonte__list', {
                    'orizzonte__list--right': orientation === 'right'
                }) }
            >
                { this.renderItems() }
                { this.renderDoneBtn() }
            </ul>
        );
    }
}

List.propTypes = {
    doneBtn: PropTypes.bool,
    doneBtnLabel: PropTypes.string,
    values: PropTypes.object,
    isFilterGroup: PropTypes.bool,
    items: PropTypes.array.isRequired,
    onApply: PropTypes.func,
    onUpdate: PropTypes.func,
    orientation: PropTypes.oneOf([
        'left',
        'right'
    ])
};

List.defaultProps = {
    doneBtn: true,
    doneBtnLabel: null,
    values: {},
    isFilterGroup: false,
    onApply: () => {},
    onUpdate: () => {},
    orientation: 'left'
};

export default List;
