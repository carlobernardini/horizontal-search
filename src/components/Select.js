import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/Filter.scss';
import '../scss/Select.scss';

const Select = ({ disabled, label, onUpdate, options, value }) => (
    <div
        className="orizzonte__filter"
    >
        <div
            className="orizzonte__filter-caption"
        >
            { label }
        </div>
        <select
            className={ classNames('orizzonte__filter-select', {
                'orizzonte__filter-select--disabled': disabled
            }) }
            disabled={ disabled }
            onChange={ (e) => {
                const { value: val } = e.target;
                onUpdate(val);
            }}
            value={ value }
        >
            { options.map((option, i) => (
                <option
                    key={ `${ option.value }.${ i }` }
                    value={ option.value }
                >
                    { option.label || option.value }
                </option>
            )) }
        </select>
    </div>
);

Select.propTypes = {
    /** If the select should be disabled */
    disabled: PropTypes.bool,
    /** Label for this filter section */
    label: PropTypes.string.isRequired,
    /** Internal callback for when select value has changed */
    onUpdate: PropTypes.func,
    /** List of selectable options (value is required) */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string,
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired
        })
    ).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string
    ])
};

Select.defaultProps = {
    disabled: false,
    onUpdate: () => {},
    value: null
};

export default Select;
