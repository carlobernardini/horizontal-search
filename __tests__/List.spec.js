import React from 'react';
import Choices from '../src/components/Choices';
import List from '../src/components/List';
import Select from '../src/components/Select';

describe('<List />', () => {
    it('should render a list of filters', () => {
        const fieldName = 'myAPIField';
        const onApply = jest.fn();
        const onUpdate = jest.fn();
        const wrapper = shallow(
            <List
                isFilterGroup
                onApply={ onApply }
                onUpdate={ onUpdate }
                items={ [
                    <Select
                        fieldName={ fieldName }
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3
                        }]}
                        notSetLabel="None"
                    />,
                    <Choices
                        fieldName="myAPIField"
                        label="Test multiple choices"
                        options={ [{
                            label: 'Test value 1',
                            value: 1,
                            disabled: true
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3,
                            disabled: true
                        }]}
                        multiple
                    />
                ]}
            />
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.find(Select).props().onUpdate(1);
        expect(onUpdate).toHaveBeenCalledWith(fieldName, 1);

        wrapper.find('.orizzonte__list-done').simulate('click');
        expect(onApply).toHaveBeenCalled();
    });

    it('should render a list of filters without done button', () => {
        const wrapper = shallow(
            <List
                isFilterGroup
                doneBtn={ false }
                items={ [
                    <Select
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3
                        }]}
                        onUpdate={() => {}}
                    />,
                    <Choices
                        fieldName="myAPIField"
                        label="Test multiple choices"
                        options={ [{
                            label: 'Test value 1',
                            value: 1,
                            disabled: true
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3,
                            disabled: true
                        }]}
                        onUpdate={() => {}}
                        multiple
                    />
                ]}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a list of filters with clear button, custom done button and custom orientation', () => {
        const onClear = jest.fn();

        const wrapper = shallow(
            <List
                isFilterGroup
                clearBtn
                doneBtnLabel="Apply"
                orientation="right"
                onClear={ onClear }
                items={ [
                    <Select
                        label="Test select"
                        options={ [{
                            label: 'Test value 1',
                            value: 1
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3
                        }]}
                        onUpdate={() => {}}
                    />,
                    <Choices
                        fieldName="myAPIField"
                        label="Test multiple choices"
                        options={ [{
                            label: 'Test value 1',
                            value: 1,
                            disabled: true
                        }, {
                            label: 'Test value 2',
                            value: 2
                        }, {
                            label: 'Test value 3',
                            value: 3,
                            disabled: true
                        }]}
                        onUpdate={() => {}}
                        multiple
                    />
                ]}
            />
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.find('.orizzonte__list-clear').simulate('click');
        expect(onClear).toHaveBeenCalled();
    });

    it('should render a list of clickable items', () => {
        const items = [{
            label: 'Filter 1'
        }, {
            label: 'Filter 2'
        }];

        const wrapper = shallow(
            <List
                items={ items.map((item) => (
                    // eslint-disable-next-line jsx-a11y/anchor-is-valid
                    <a
                        href="#"
                        className="orizzonte__item-clickable"
                        onClick={ jest.fn() }
                    >
                        { item.label }
                    </a>
                ))}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
