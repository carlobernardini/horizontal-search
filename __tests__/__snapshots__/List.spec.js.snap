// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render a list of clickable items 1`] = `
<ul
  className="orizzonte__list"
  style={
    Object {
      "minWidth": null,
    }
  }
>
  <li
    className="orizzonte__item"
    key="0/.$Filter1"
  >
    <a
      className="orizzonte__item-clickable"
      href="#"
      key="Filter1"
      onClick={[MockFunction]}
    >
      Filter 1
    </a>
  </li>
  <li
    className="orizzonte__item"
    key="1/.$Filter2"
  >
    <a
      className="orizzonte__item-clickable"
      href="#"
      key="Filter2"
      onClick={[MockFunction]}
    >
      Filter 2
    </a>
  </li>
</ul>
`;

exports[`<List /> should render a list of filters 1`] = `
<ul
  className="orizzonte__list"
  style={
    Object {
      "minWidth": null,
    }
  }
>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="0/.0"
  >
    <OrizzonteSelect
      disabled={false}
      fieldName="myAPIField"
      information={null}
      label="Test select"
      notSetLabel="None"
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="1/.1"
  >
    <OrizzonteChoices
      fieldName="myAPIField"
      information={null}
      label="Test multiple choices"
      multiple={true}
      noPreferenceLabel={null}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "disabled": true,
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "disabled": true,
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
  <li
    className="orizzonte__list-controls"
  >
    <button
      className="orizzonte__list-control orizzonte__list-done"
      onClick={[MockFunction]}
      type="button"
    >
      Done
    </button>
  </li>
</ul>
`;

exports[`<List /> should render a list of filters with remove button, custom done button and custom orientation 1`] = `
<ul
  className="orizzonte__list orizzonte__list--right"
  style={
    Object {
      "minWidth": null,
    }
  }
>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="0/.0"
  >
    <OrizzonteSelect
      disabled={false}
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="1/.1"
  >
    <OrizzonteChoices
      fieldName="myAPIField"
      information={null}
      label="Test multiple choices"
      multiple={true}
      noPreferenceLabel={null}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "disabled": true,
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "disabled": true,
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
  <li
    className="orizzonte__list-controls"
  >
    <button
      className="orizzonte__list-control orizzonte__list-remove"
      onClick={[MockFunction]}
      type="button"
    >
      Remove group
    </button>
    <button
      className="orizzonte__list-control orizzonte__list-done"
      onClick={[Function]}
      type="button"
    >
      Apply
    </button>
  </li>
</ul>
`;

exports[`<List /> should render a list of filters without done button 1`] = `
<ul
  className="orizzonte__list"
  style={
    Object {
      "minWidth": null,
    }
  }
>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="0/.0"
  >
    <OrizzonteSelect
      disabled={false}
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
  <li
    className="orizzonte__item orizzonte__item--filters"
    key="1/.1"
  >
    <OrizzonteChoices
      fieldName="myAPIField"
      information={null}
      label="Test multiple choices"
      multiple={true}
      noPreferenceLabel={null}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "disabled": true,
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
          Object {
            "disabled": true,
            "label": "Test value 3",
            "value": 3,
          },
        ]
      }
      value={null}
    />
  </li>
</ul>
`;
